src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs

```diff
@@ -199,37 +200,122 @@ public BuildContractTransactionResult BuildTx(BuildContractTransactionRequest re
             return BuildContractTransactionResult.Success(model);
         }
 
-        private string[] ReplaceSignatures(string[] parameters, string[] signatures)
+        /// <summary>
+        /// Parses elements until the ',' (separator) or ']' (ending) special character is encountered.
+        /// Single quotes are used to enclose text regions to preserve whitespace or to make ']' or ',' non-special.
+        /// </summary>
+        /// <param name="parameter">The parameter to parse.</param>
+        /// <param name="position">Initally the position in the string following the '[' character. This gets incremented as we traverse the parameter.</param>
+        /// <returns>The elements as a string array.</returns>
+        /// <remarks>
+        /// Single quotes are removed from elements and double single quotes are converted to literal single quotes.
+        /// </remarks>
+        public static string[] ParseArray(string parameter, ref int position)
```

Comment:
```
        public static string[] ParseArray(string parameter, int position)
        {
            var elements = new List<string>();
            int elementStart = position;
            bool quoted = false;

            for (; position < parameter.Length; position++)
            {
                if (parameter[position] == '\'')
                {
                    quoted = !quoted;
                }
                else if (!quoted && (parameter[position] == ',' || parameter[position] == ']'))
                {
                    string placeHolder = "\x0";

                    // Extract the element.
                    // Replace double-single-quotes with a placeholder. Remove single-quotes. Replace the placeholder with a single-quote.
                    string element = parameter.Substring(elementStart, position - elementStart).Trim().Replace("''", placeHolder).Replace("'", "").Replace(placeHolder, "'");
                    elements.Add(element);
                    if (parameter[position] == ']')
                        return elements.ToArray();
                    elementStart = position + 1;
                }
            }

            throw new Exception($"Parameter '{parameter}' has an invalid array syntax at character {position}.");
        }
```

Reply:
Thanks @YakupIpek . The reason for using `ref` is that it allows `ParseArray` to be part of a bigger parsing experience where there may be more characters following the `[` so I want to stick with the current implementation in this case.